Quotient/events/roommessageevent.h

```diff
@@ -62,49 +62,37 @@ class QUOTIENT_API RoomMessageEvent : public RoomEvent {
     //! Update the message JSON with the given content
     void setContent(std::unique_ptr<EventContent::Base> content);
 
-    QMimeType mimeType() const;
-
-    //! \brief Determine whether the message has text content
-    //!
-    //! \return true, if the message type is one of m.text, m.notice, m.emote,
-    //!         or the message type is unspecified (in which case plainBody()
-    //!         can still be examined); false otherwise
-    bool hasTextContent() const;
-
-    //! \brief Get the TextContent object for the event
+    //! \brief Determine whether the message has content/attachment of a specified type
     //!
-    //! \return A TextContent object if the message has one; std::nullopt otherwise.
-    std::unique_ptr<EventContent::TextContent> richTextContent() const;
+    //! \return true, if the message has type and content corresponding to \p ContentT (e.g.
+    //!         `m.file` or `m.audio` for FileContent); false otherwise
+    template <std::derived_from<EventContent::Base> ContentT>
+    bool has() const { return false; }
 
-    //! \brief Determine whether the message has a file/attachment
+    //! \brief Get the message content and try to cast it to the specified type
     //!
-    //! \return true, if the message has a data structure corresponding to
-    //!         a file (such as m.file or m.audio); false otherwise
-    bool hasFileContent() const;
+    //! \return A pointer to the object of the requested type if the event has content of this type;
+    //!         nullptr, if the event has no content or the content cannot be cast to this type
+    template <std::derived_from<EventContent::Base> ContentT>
+    std::unique_ptr<ContentT> get() const
+    {
+        return has<ContentT>()
+                   ? std::unique_ptr<ContentT>(static_cast<ContentT*>(content().release()))
+                   : nullptr;
+    }
```

Comment:
So this leads to a nullptr being returned if I ask for `FileContent` on and Image.

The reason this caught me out is that in NeoChat when getting media info to show in the UI it's done in 2 steps:
1.  Grab all the generic params (by grabbing a `FileContent` currently
2. Cast to specific type to get the unique ones.

Now it wouldn't be hard to change to make this work (get the specialised content object straight away) but it still feels weird that `FileContentBase` doesn't have all the generic params present in each filetype e.g. source, filename

Reply:
Well, yes, `FileContentBase` is effectively a virtual API for `FileInfo` that only happens to expose `url()` at the moment. I wanted to go with as limited set as possible because I still hope to throw away the whole thing and replace it with something easier to consume (I have some thoughts around the new hierarchy but they are rather disorganised).

I can probably add `has<FileInfo>()` and `get<FileInfo>()` et al. as well - probably replacing `has/get` for `FileContentBase` as it exposes a subset of `FileInfo`. `FileInfo` is not a polymorphic class though - you cannot call `toJson()` on `std::unique_ptr<FileInfo>`, for one - and I remember a place in Quaternion that uses `FileContentBase` being derived from `Base`. So probably I'll have to keep all three, although it's very confusing.

Reply:
As I said I can work round it so unless we add just some simple virtual members to `FileConentBase` it's probably not worth it right now. Just document it clearly that you'll get a null from `get<FileConent>` or other if the file content isn't an exact match e.g. `m.file` for `FileContent`

Reply:
I played around with this. First of all, the very similar concern to yours holds even for the library itself, so fixing this is absolutely warranted. And I think I found a reasonable middle ground - `FileContentBase::commonInfo()` will return a _copy_ of `FileInfo` to handle the NeoChat's case, and it also now has `Thumbnail` so you can show a thumbnail for any file content without even figuring out the type of that content. There are still cases where this is not enough (e.g. updating the file info cannot easily be done; I had to add `RoomMessageEvent::updateFileSource()` to support a pending event case) but I hope we can stop at that for now without affecting the code base further. Commits coming rsn.